// marko/dist/taglib/index.js
"use strict";

// eslint-disable-next-line no-constant-binary-expression

const compiler = require("@marko/compiler");

function clearCache() {
  compiler.taglib.clearCaches();
}

function register(taglibProps, taglibPath) {
  return compiler.taglib.register(taglibPath, taglibProps);
}

function registerFromFile(taglibPath) {
  return register(
    compiler.taglib._loader.loadTaglibFromFile(taglibPath),
    taglibPath
  );
}

exports.clearCache = clearCache;
exports.register = register;
exports.registerFromFile = registerFromFile;
exports.buildLookup = compiler.taglib.buildLookup;
exports.excludeDir = compiler.taglib.excludeDir;
exports.excludePackage = compiler.taglib.excludePackage;
exports.loader = compiler.taglib._loader;
exports.finder = compiler.taglib._finder;
exports.lookup = {
  buildLookup: function (dir, translator) {
    if (!translator || !Array.isArray(translator.taglibs)) {
      translator = require("@marko/translator-default");
      // eslint-disable-next-line no-constant-condition





    }

    return compiler.taglib.buildLookup(dir, translator);
  },
  registerTaglib(taglib) {
    if (typeof taglib === "string") {
      registerFromFile(taglib);
    } else {
      register(taglib, taglib.id);
    }
  }
};
