// @rspack/core/dist/Module.js
"use strict";
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var _ContextModuleFactoryBeforeResolveData_inner, _ContextModuleFactoryAfterResolveData_inner, _Module_inner, _Module_originalSource, _CodeGenerationResult_inner;
Object.defineProperty(exports, "__esModule", { value: true });
exports.CodeGenerationResults = exports.CodeGenerationResult = exports.Module = exports.ContextModuleFactoryAfterResolveData = exports.ContextModuleFactoryBeforeResolveData = void 0;
const DependenciesBlock_1 = require("./DependenciesBlock");
const Dependency_1 = require("./Dependency");
const source_1 = require("./util/source");
class ContextModuleFactoryBeforeResolveData {
    static __from_binding(binding) {
        return new ContextModuleFactoryBeforeResolveData(binding);
    }
    static __to_binding(data) {
        return __classPrivateFieldGet(data, _ContextModuleFactoryBeforeResolveData_inner, "f");
    }
    constructor(binding) {
        _ContextModuleFactoryBeforeResolveData_inner.set(this, void 0);
        __classPrivateFieldSet(this, _ContextModuleFactoryBeforeResolveData_inner, binding, "f");
        Object.defineProperties(this, {
            context: {
                enumerable: true,
                get() {
                    return binding.context;
                },
                set(val) {
                    binding.context = val;
                }
            },
            request: {
                enumerable: true,
                get() {
                    return binding.request;
                },
                set(val) {
                    binding.request = val;
                }
            },
            regExp: {
                enumerable: true,
                get() {
                    if (!binding.regExp) {
                        return undefined;
                    }
                    const { source, flags } = binding.regExp;
                    return new RegExp(source, flags);
                },
                set(val) {
                    if (!val) {
                        binding.regExp = undefined;
                        return;
                    }
                    binding.regExp = {
                        source: val.source,
                        flags: val.flags
                    };
                }
            },
            recursive: {
                enumerable: true,
                get() {
                    return binding.recursive;
                },
                set(val) {
                    binding.recursive = val;
                }
            }
        });
    }
}
exports.ContextModuleFactoryBeforeResolveData = ContextModuleFactoryBeforeResolveData;
_ContextModuleFactoryBeforeResolveData_inner = new WeakMap();
class ContextModuleFactoryAfterResolveData {
    static __from_binding(binding) {
        return new ContextModuleFactoryAfterResolveData(binding);
    }
    static __to_binding(data) {
        return __classPrivateFieldGet(data, _ContextModuleFactoryAfterResolveData_inner, "f");
    }
    constructor(binding) {
        _ContextModuleFactoryAfterResolveData_inner.set(this, void 0);
        __classPrivateFieldSet(this, _ContextModuleFactoryAfterResolveData_inner, binding, "f");
        Object.defineProperties(this, {
            resource: {
                enumerable: true,
                get() {
                    return binding.resource;
                },
                set(val) {
                    binding.resource = val;
                }
            },
            context: {
                enumerable: true,
                get() {
                    return binding.context;
                },
                set(val) {
                    binding.context = val;
                }
            },
            request: {
                enumerable: true,
                get() {
                    return binding.request;
                },
                set(val) {
                    binding.request = val;
                }
            },
            regExp: {
                enumerable: true,
                get() {
                    if (!binding.regExp) {
                        return undefined;
                    }
                    const { source, flags } = binding.regExp;
                    return new RegExp(source, flags);
                },
                set(val) {
                    if (!val) {
                        binding.regExp = undefined;
                        return;
                    }
                    binding.regExp = {
                        source: val.source,
                        flags: val.flags
                    };
                }
            },
            recursive: {
                enumerable: true,
                get() {
                    return binding.recursive;
                },
                set(val) {
                    binding.recursive = val;
                }
            },
            dependencies: {
                enumerable: true,
                get() {
                    return binding.dependencies.map(dep => Dependency_1.Dependency.__from_binding(dep));
                }
            }
        });
    }
}
exports.ContextModuleFactoryAfterResolveData = ContextModuleFactoryAfterResolveData;
_ContextModuleFactoryAfterResolveData_inner = new WeakMap();
class Module {
    static __from_binding(module, compilation) {
        return new Module(module, compilation);
    }
    constructor(module, compilation) {
        _Module_inner.set(this, void 0);
        _Module_originalSource.set(this, void 0);
        __classPrivateFieldSet(this, _Module_inner, module, "f");
        this.type = module.type;
        this.layer = module.layer ?? null;
        this.context = module.context;
        this.resource = module.resource;
        this.request = module.request;
        this.userRequest = module.userRequest;
        this.rawRequest = module.rawRequest;
        this.factoryMeta = module.factoryMeta;
        const customModule = compilation?.__internal__getCustomModule(module.moduleIdentifier);
        this.buildInfo = customModule?.buildInfo || {};
        this.buildMeta = customModule?.buildMeta || {};
    }
    originalSource() {
        if (__classPrivateFieldGet(this, _Module_originalSource, "f"))
            return __classPrivateFieldGet(this, _Module_originalSource, "f");
        if (__classPrivateFieldGet(this, _Module_inner, "f").originalSource) {
            __classPrivateFieldSet(this, _Module_originalSource, source_1.JsSource.__from_binding(__classPrivateFieldGet(this, _Module_inner, "f").originalSource), "f");
            return __classPrivateFieldGet(this, _Module_originalSource, "f");
        }
        return null;
    }
    identifier() {
        return __classPrivateFieldGet(this, _Module_inner, "f").moduleIdentifier;
    }
    nameForCondition() {
        if (typeof __classPrivateFieldGet(this, _Module_inner, "f").nameForCondition === "string") {
            return __classPrivateFieldGet(this, _Module_inner, "f").nameForCondition;
        }
        return null;
    }
    get blocks() {
        if ("blocks" in __classPrivateFieldGet(this, _Module_inner, "f")) {
            return __classPrivateFieldGet(this, _Module_inner, "f").blocks.map(b => new DependenciesBlock_1.DependenciesBlock(b));
        }
        return [];
    }
    size(type) {
        if ("size" in __classPrivateFieldGet(this, _Module_inner, "f")) {
            return __classPrivateFieldGet(this, _Module_inner, "f").size(type);
        }
        return 0;
    }
}
exports.Module = Module;
_Module_inner = new WeakMap(), _Module_originalSource = new WeakMap();
class CodeGenerationResult {
    constructor(result) {
        _CodeGenerationResult_inner.set(this, void 0);
        __classPrivateFieldSet(this, _CodeGenerationResult_inner, result, "f");
    }
    get(sourceType) {
        return __classPrivateFieldGet(this, _CodeGenerationResult_inner, "f").sources[sourceType];
    }
}
exports.CodeGenerationResult = CodeGenerationResult;
_CodeGenerationResult_inner = new WeakMap();
class CodeGenerationResults {
}
exports.CodeGenerationResults = CodeGenerationResults;
