// @rspack/core/dist/RuntimeModule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RuntimeModule = exports.RuntimeModuleStage = void 0;
var RuntimeModuleStage;
(function (RuntimeModuleStage) {
    RuntimeModuleStage[RuntimeModuleStage["NORMAL"] = 0] = "NORMAL";
    RuntimeModuleStage[RuntimeModuleStage["BASIC"] = 5] = "BASIC";
    RuntimeModuleStage[RuntimeModuleStage["ATTACH"] = 10] = "ATTACH";
    RuntimeModuleStage[RuntimeModuleStage["TRIGGER"] = 20] = "TRIGGER";
})(RuntimeModuleStage = exports.RuntimeModuleStage || (exports.RuntimeModuleStage = {}));
class RuntimeModule {
    static __to_binding(compilation, module) {
        return {
            name: module.name,
            stage: module.stage,
            generator: module.generate.bind(module),
            cacheable: !(module.fullHash || module.dependentHash),
            isolate: module.shouldIsolate()
        };
    }
    constructor(name, stage = RuntimeModuleStage.NORMAL) {
        this.fullHash = false;
        this.dependentHash = false;
        this.chunk = null;
        this.compilation = null;
        this.chunkGraph = null;
        this._name = name;
        this._stage = stage;
    }
    attach(compilation, chunk, chunkGraph) {
        this.compilation = compilation;
        this.chunk = chunk;
        this.chunkGraph = chunkGraph;
    }
    get name() {
        return this._name;
    }
    get stage() {
        return this._stage;
    }
    identifier() {
        return `webpack/runtime/${this._name}`;
    }
    readableIdentifier() {
        return `webpack/runtime/${this._name}`;
    }
    shouldIsolate() {
        return true;
    }
    generate() {
        throw new Error(`Should implement "generate" method of runtime module "${this.name}"`);
    }
}
RuntimeModule.STAGE_NORMAL = RuntimeModuleStage.NORMAL;
RuntimeModule.STAGE_BASIC = RuntimeModuleStage.BASIC;
RuntimeModule.STAGE_ATTACH = RuntimeModuleStage.ATTACH;
RuntimeModule.STAGE_TRIGGER = RuntimeModuleStage.TRIGGER;
exports.RuntimeModule = RuntimeModule;
