// @marko/express/dist/index.mjs
// src/redirect.ts
import { ServerResponse } from "http";
var kCSPNonce = Symbol("cspNonce");
var kOnWriteFlush = Symbol("on-flush");
function redirectWithMidstreamSupport(p1, p2) {
  var _a;
  const status = typeof p1 === "number" ? p1 : 302;
  const redirectUrl = typeof p1 === "string" ? p1 : p2;
  const nonce = this[kCSPNonce] ? ` nonce=${JSON.stringify(this[kCSPNonce])}` : "";
  if (this.headersSent && ((_a = this.getHeader("Content-Type")) == null ? void 0 : _a.startsWith("text/html"))) {
    this[kOnWriteFlush] = this.destroy;
    this.write(`
        <meta http-equiv=refresh content=${JSON.stringify(`0;url=${redirectUrl}`)}>
        <script${nonce}>location.href=${JSON.stringify(redirectUrl)}<\/script>
      `);
    if (this.flush)
      this.flush();
    this.write = noopTrue;
    this.flush = this.end = noopThis;
  } else {
    Object.getPrototypeOf(this).redirect.call(this, status, redirectUrl);
  }
}
function noopThis() {
  return this;
}
function noopTrue() {
  return true;
}
var _write = ServerResponse.prototype.write;
ServerResponse.prototype.write = function(data, encoding, callback) {
  if (!this[kOnWriteFlush])
    return _write.call(this, data, encoding, callback);
  if (typeof encoding === "function") {
    callback = encoding;
    encoding = "utf8";
  }
  return _write.call(this, data, encoding, () => {
    this[kOnWriteFlush]();
    callback == null ? void 0 : callback();
  });
};

// src/index.ts
function middleware() {
  return (_req, res, next) => {
    res.marko = renderMarkoTemplate;
    res.redirect = redirectWithMidstreamSupport;
    next();
  };
}
function renderMarkoTemplate(template, input) {
  const $global = { ...this.app.locals, ...this.locals };
  if (input) {
    if (input.$global) {
      Object.assign($global, input.$global);
    }
    input.$global = $global;
  }
  this[kCSPNonce] = $global.cspNonce;
  this.set("Content-Type", "text/html; charset=utf-8");
  template.render(input || { $global }, this).on("error", this.req.next);
}
export {
  middleware as default
};
